File: attb/app/models/attendance.py
from typing import Dict, List
from pydantic import BaseModel, Field
from app.core.constants import AttendanceStatus


class SubjectAttendance(BaseModel):
    """Attendance statistics for a single subject."""
    subject_code: str
    subject_name: str
    total_classes: int
    attended: int
    absent: int
    cancelled: int
    scheduled: int  # Future classes
    percentage: float
    status: AttendanceStatus
    classes_needed_for_75: int  # How many more to attend to reach 75%
    classes_can_miss: int  # How many can be missed while staying above 75%
    
    class Config:
        use_enum_values = True


class OverallAttendance(BaseModel):
    """Overall attendance summary across all subjects."""
    total_classes: int
    attended: int
    absent: int
    cancelled: int
    scheduled: int
    overall_percentage: float
    status: AttendanceStatus
    subjects: List[SubjectAttendance]
    
    class Config:
        use_enum_values = True


class AttendanceWarning(BaseModel):
    """Warning message for attendance issues."""
    subject_code: str
    subject_name: str
    current_percentage: float
    message: str
    severity: str  # "info", "warning", "critical"
    recommendation: str


class AttendanceProjection(BaseModel):
    """Projected attendance after simulated changes."""
    subject_code: str
    current_percentage: float
    projected_percentage: float
    change: float
    classes_simulated: int
    will_meet_threshold: bool

File: att/src/types/attendance.ts
// Attendance types
export type AttendanceStatus = 'present' | 'absent' | 'planned-present' | 'planned-absent';

export interface AttendanceRecord {
  id?: string;
  date: string; // ISO date format
  subjectCode: string;
  slotNumber: number;
  status: AttendanceStatus;
  isPast: boolean; // Whether the date is in the past
}

export interface SubjectAttendance {
  subjectCode: string;
  subjectName: string;
  totalClasses: number;
  attendedClasses: number;
  plannedPresentClasses: number;
  plannedAbsentClasses: number;
  currentPercentage: number;
  projectedPercentage: number;
  isBelowThreshold: boolean;
  willBeBelowThreshold: boolean;
}

export interface AttendanceSummary {
  subjects: SubjectAttendance[];
  overallPercentage: number;
  totalClasses: number;
  totalAttended: number;
}

export interface AttendanceWarning {
  subjectCode: string;
  subjectName: string;
  currentPercentage: number;
  projectedPercentage: number;
  message: string;
  severity: 'critical' | 'warning' | 'info';
}
